### PR TITLE
Cleanup DataSource initialization [DPP-583]

### PR DESCRIPTION
**Pull request by nmarton-da**
* Introduce VerifiedDataSource to capture async verification and factoring
* Move compatibility check to VerifiedDataSource
* Wire to FlywayMigrations
* Cleanup HikariDataSourceOwner

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/VerifiedDataSource.scala [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import com.daml.logging.{ContextualizedLogger, LoggingContext}
+import com.daml.platform.store.DbType
+import com.daml.timer.RetryStrategy
+import javax.sql.DataSource
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.concurrent.duration.DurationInt
+import scala.util.Using
+
+object VerifiedDataSource {
```
Comment (rautenrieth-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/backend/VerifiedDataSource.scala:15]: There are many classes in `com.daml.platform.store.backend`. At least for me, a short documentation of each class helps me navigating the code base.

```suggestion
/** Returns a DataSource that is guaranteed to be connected to a responsive, compatible database. */
object VerifiedDataSource {
```

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/common/CommonStorageBackend.scala [hunk 2]
```diff
@@ -705,4 +705,7 @@ private[backend] trait CommonStorageBackend[DB_BATCH] extends StorageBackend[DB_
       stmnt.close()
     }
   }
+
+  override def checkDatabaseAvailable(connection: Connection): Unit =
+    assert(SQL"SELECT 1".as(get[Int](1).single)(connection) == 1)
```
Comment (rautenrieth-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/backend/common/CommonStorageBackend.scala:710]: What log output will this produce if the database is not available? The output of the `assert` doesn't look useful. If if even can fail at all - when would `SELECT 1` not return `1`?

**Review by rautenrieth-da — APPROVED** [review]
Change looks good, left two minor comments.

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/common/CommonStorageBackend.scala (hunk 2 above)
Comment (nmarton-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/backend/common/CommonStorageBackend.scala:710]: log: if this fails the VerifiedDataSource will produce logs, that it attempts to connect to the db
select 1: with this we can make sure a full roundtrip between db and app, like a sanity check: testing all layers. I'd like to keep  it in if you don't mind.

**Review by kamil-da — APPROVED** [review]

**Pull request merged**